pyproject.toml

```diff
@@ -1,6 +1,61 @@
 [build-system]
-requires = ["setuptools", "wheel"]
-build-backend = "setuptools.build_meta"
+requires = ["hatchling>=0.14.0"]
```

Comment:
Why not flit if we're going down this path? I know that has a better story for OS repackaging.

Reply:
By OS repackaging, do you simply mean the availability of a system package for different Linux distributions like Debian? If so, as you might have seen in that other thread there is one for [Debian](https://salsa.debian.org/python-team/packages/hatchling) and the Fedora one is [nearly complete](https://github.com/ofek/hatch/issues/120#issuecomment-1043048369). For others like CentOS ([7 & 8-stream](http://mirror.centos.org/centos/7/os/x86_64/Packages/) / [9-stream](http://mirror.stream.centos.org/9-stream/AppStream/x86_64/os/Packages/)) I don't see `pipx` nor `flit`.

I appreciate Flit, and what it has accomplished. It essentially was the shining example of PEP 517, illustrating that `setuptools` was no longer a requirement for pure Python projects!

In many cases though, projects need more, which is where Hatchling comes in.

For example, many popular projects (including some you maintain 🙂) require [setuptools_scm](https://github.com/pypa/setuptools_scm) to use VCS tags exclusively for the project version. Hatchling has [a plugin](https://github.com/ofek/hatch-vcs) for this (despite the legacy naming, Ronny has taken great care to decouple `setuptools_scm` from `setuptools`). `pipx` may want this functionality in the future.

Or, if a project wanted to use [Mypyc](https://github.com/mypyc/mypyc) to compile code with the help of type hints, there is [a plugin](https://github.com/ofek/hatch-mypyc) for that too.

Or, if it is desired that the project page on PyPI displays the README and changelog (you'll often see them concatenated in `setup.py`), a custom [metadata hook](https://ofek.dev/hatch/latest/plugins/metadata-hook/) can easily be written to achieve this.

Essentially, I've put much effort into ensuring arbitrary use cases are easily supported.

Reply:
> going ultra-simple and sticking to a pypa tool

1. Both use PEP 621 metadata so the "simplicity" for metadata is identical. Defining the contents of source distributions in Flit also requires [its own section](https://flit.readthedocs.io/en/latest/pyproject_toml.html#sdist-section). Finally, Flit does not currently support the layout of `pipx` because there is a strict expectation of a variable `__version__` at the package root:

   ```
   ofek@ozone ~\Desktop\code\pipx modernize-metadata $ python -m build
   ...
       raise NoVersionError('Cannot package module without a version string. '
   flit_core.common.NoVersionError: Cannot package module without a version string. Please define a `__version__ = "x.y.z"` in your module.
   ```
   If support is implemented, it would also require [its own section](https://github.com/pypa/flit/pull/382#issuecomment-874723398).
2. I get the sense that since Flit now lives under pypa/ it is being recommended here as the preferred or standard tool. If that is indeed what's happening (correct me if I misinterpreted), then I guess that's the future of Python packaging and I'll have to accept that 🙂 Though, that goes against everything in the discussion about the transfer: https://discuss.python.org/t/should-flit-become-a-pypa-project/11854

   > If PyPA did adopt Flit, I’d like to avoid any suggestion that it’s the ‘official’ or ‘standard’ tool for creating & publishing packages.

   Perhaps to be recognized as legitimate, backends should first apply/ask to be included in PyPA? Based on:

   > I think it would be good to have multiple backends under the PyPA umbrella, _precisely_ to avoid giving any impression that there’s a “PyPA preferred” option.

Reply:
No, but _this package is a PyPA project_. If something horrible breaks, we have a little more of a say in fixing it. There are also other PyPA projects planning on moving to Flit this year.

Reply:
I see. Should we document the fact that PyPA projects must use a PyPA build backend somewhere?

Reply:
And there are a lot of repercussions for distro packagers that don't support PEP 517 - we are already having to convince them to include flit, also convincing them to include hatchling is likely not the correct approach. The correct approach is for them to use installer, but that's not ready yet. Maybe helping with that would really help out all PEP 517 backends (at least with pure wheels). See https://github.com/pypa/wheel/pull/436

conda-forge doesn't include hatch/hatchling, so this would immediately break the conda-forge recipe, while flit is available there. Etc.

Reply:
I am not saying that hatchling cannot be used by PyPA projects. I'm just giving an opinion and warning that at [brand new](https://pypistats.org/packages/hatchling) backend with seven dependencies is likely to be more risky than using something that has been around for years, has been involved in the PEP process, was popular enough to make it into the PyPA, and has zero dependencies. Maintaining a backend is hard, and a packages is only as compatible as its backend. There is a _lot_ involved, like PyPy support, and having lots of users and lots of experience working through all the exotic situations is quite important. I wish Hatchling the best, and was happy to add support for it to scikit-hep/cookie, but "it supports PEP 621" doesn't mean you can throw out the rest of the considerations of picking a good dependency. Flit has the added bonus of being "in house" for another PyPA project, and several projects are planning on going through the pain of conversion together. I'm quite sure Flit will continue to support new Python versions, new Python distributions (like Pyodide), etc. Maybe Hatch will too, but I'd like to see with more than 10 days of heavy use.

I am not a pipx maintainer, I do not have any say, I'm just providing some advice. The pipx maintainers might happy use Hatchling after reading this advice, and I would be happy for them. Though if it breaks GitHub Action's recipe, I'll be furious. ;)

Just curious, do you know why hatchling suddenly started getting more than 100 downloads per day 10 days ago? Did a major project pick it up?

Reply:
> Just curious, do you know why hatchling suddenly started getting more than 100 downloads per day 10 days ago? Did a major project pick it up?

Yes, after I deemed it stable (esp. config options) I made an RFC internally proposing modernizing packaging (in particular for integrations) and we chose Hatchling over the other backends. As mentioned in the OP, this repo is almost done switching: https://github.com/DataDog/integrations-core

After the integration repos then I'll do the other Python projects we own, though those are mostly private so I cannot provide links.

Reply:
> By OS repackaging,

I've primarily been referring to ease of bootstrapping hatch within a Linux distribution 🤔

Reply:
I'd be happy to use hatch over flit if you could fix the Linux distribution bootstrap problem for Arch/RHEL/Debian maintainers.

Reply:
For reference, here's the plan for build: https://github.com/pypa/build/issues/394#issuecomment-1044550493

Don't forget conda-forge, WebAssembly efforts like Pyodide, etc.

Reply:
pipx isn't a bootstrapping dependency, not sure why that would matter here.

Reply:
> As in vendoring or is there some other way?

I don't think arbitrary backends need to support bootstrapping, and very much should not be vendoring.

Reply:
I think I got confused here, as long as flit is bootstrap-able (aka has no dependencies) and build project is using build, any distribution can use build + pip/installer to generate a OS package for any other build backed (including hatch).

Reply:
The original request was: “One other thing I've been wanting to do for a while is bring back the install script where you can run `curl ... | python3`. That is originally how pipx worked.“ - that’s part of where talk of bootstrapping came from, I think. Though that’s not quite bootstrapping.

Anyway, installer will really help for using arbitrary backends, since then distros can use the wheel to package without ever touching the backend.